### PR TITLE
[release-tooling] Implement a parser from yaml

### PR DESCRIPTION
**Pull request by runtian-zhou**
### Description

Implement the parser to parse the release config from a yaml file. Added generation logic for consensus config as well.

### Test Plan

`cargo run -- write-default --output-path data/example.yaml`

and `cargo run -- generate-proposals --release-config  data/example.yaml --output-dir <OUTPUT_DIR>`

The upgrade smoke test will make sure upgrading consensus config would work as well.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5562)

### CODE REVIEW

File: aptos-move/aptos-release-builder/src/components/consensus_config.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::utils::*;
+use anyhow::Result;
+use aptos_types::on_chain_config::OnChainConsensusConfig;
+use move_model::{code_writer::CodeWriter, emit, emitln, model::Loc};
+
+pub fn generate_consensus_upgrade_proposal(
```
Comment (runtian-zhou) [inline aptos-move/aptos-release-builder/src/components/consensus_config.rs:9]: This logic is migrated from #5527. Maybe we could refactor that test later cc @igor-aptos

File: aptos-move/aptos-release-builder/data/example.yaml [hunk 2]
```diff
@@ -0,0 +1,472 @@
+---
```
Comment (runtian-zhou) [inline aptos-move/aptos-release-builder/data/example.yaml:1]: This is an example release config that contains all the information we could set using on chain config. Governance proposals could be generated automatically from this yaml file.

File: aptos-move/aptos-release-builder/src/components/consensus_config.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    consensus_config: &OnChainConsensusConfig,
+    is_testnet: bool,
+) -> Result<Vec<(String, String)>> {
+    let mut result = vec![];
+
+    let gas_schedule_blob = bcs::to_bytes(consensus_config).unwrap();
```
Comment (movekevin) [inline aptos-move/aptos-release-builder/src/components/consensus_config.rs:15]: nit: Wrong name. Should be consensus_config_blob.

File: aptos-move/aptos-release-builder/src/components/consensus_config.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
```
Comment (movekevin) [inline aptos-move/aptos-release-builder/src/components/consensus_config.rs:16]: nit: Unnecessary empty line.

File: aptos-move/aptos-release-builder/src/components/consensus_config.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    assert!(gas_schedule_blob.len() < 65536);
+
+    let writer = CodeWriter::new(Loc::default());
+
+    emitln!(writer, "// Consensus config upgrade proposal\n");
+
+    if is_testnet {
```
Comment (movekevin) [inline aptos-move/aptos-release-builder/src/components/consensus_config.rs:23]: nit: This is getting a bit redundant. Refactor this so each component doesn't have to do this themselves?

File: aptos-move/aptos-release-builder/src/components/feature_flags.rs [hunk 6]
```diff
@@ -3,14 +3,24 @@
...
+#[derive(Clone, Deserialize, PartialEq, Eq, Serialize)]
 pub struct Features {
     pub enabled: Vec<FeatureFlag>,
     pub disabled: Vec<FeatureFlag>,
 }
 
+#[derive(Clone, Deserialize, PartialEq, Eq, Serialize)]
+#[allow(non_camel_case_types)]
+#[serde(rename_all = "snake_case")]
+pub enum FeatureFlag {
```
Comment (movekevin) [inline aptos-move/aptos-release-builder/src/components/feature_flags.rs:19]: So every time we add a new feature flag to the yaml file, we'd need to update this? I don't think it's a problem - just confirming.

File: aptos-move/aptos-release-builder/src/components/mod.rs [hunk 7]
```diff
@@ -59,14 +79,54 @@ impl ReleaseConfig {
...
+        Ok(())
+    }
+
+    pub fn parse(serialized: &str) -> Result<Self> {
+        serde_yaml::from_str(serialized).map_err(|e| anyhow!("Failed to parse the config: {:?}", e))
+    }
 }
 
 impl Default for ReleaseConfig {
     fn default() -> Self {
```
Comment (movekevin) [inline aptos-move/aptos-release-builder/src/components/mod.rs:122]: Can you remind me where this is used again?

File: aptos-move/aptos-release-builder/src/components/feature_flags.rs (hunk 6 above)
Comment (runtian-zhou) [reply aptos-move/aptos-release-builder/src/components/feature_flags.rs:19]: This is correct. I'm expecting us to update this.

File: aptos-move/aptos-release-builder/src/components/consensus_config.rs (hunk 5 above)
Comment (runtian-zhou) [reply aptos-move/aptos-release-builder/src/components/consensus_config.rs:23]: Yea I think I can refactor a bit here.

File: aptos-move/aptos-release-builder/src/components/mod.rs (hunk 7 above)
Comment (runtian-zhou) [reply aptos-move/aptos-release-builder/src/components/mod.rs:122]: This is used in the upgrade test to make sure that the proposal generation process would always work. Though in the future we should be replacing this with the real release config we want to put for the next release cycle.

Comment (movekevin) [reply aptos-move/aptos-release-builder/src/components/mod.rs:122]: "in the future we should be replacing this with the real release config we want to put for the next release cycle." => What do you mean? Isn't this the default only?

**Review by zekun000 — COMMENTED** [review]
oh this is really nice to have

**Comment by runtian-zhou** [comment]
Updated the PR with the refactor. Tho it doesn't look cleaner than before either... lemme know if you would like to keep it or not cc @movekevin

File: aptos-move/aptos-release-builder/src/components/mod.rs (hunk 7 above)
Comment (runtian-zhou) [reply aptos-move/aptos-release-builder/src/components/mod.rs:122]: Sry I mean that right now, we use `ReleaseConfig::default` in the testing pipeline. We should expect to see another test that uses sth like `ReleaseConfig::upcoming_release` to make sure the config is indeed valid. Hope that makes sense.

**Comment by igor-aptos** [comment]
This is probably beyond this PR, but what is the expected flow?

It seems to me release yaml is already materialized release decision - delta between two versions, i.e. what needs to be upgraded, for a specific network (i.e. testnet or not). Correct? 
How are we going to generate all release yamls for a single release, for all the networks we have? (devnet, testnet, mainnet)

I.e. how would we know if any configs have changed, and whether governance proposal for updating onchain config (for example consensus one) is going to be needed?

Are we going to have a separate "snapshot" files telling us the state of various networks, and then create a snapshot for new release, and then create there release yamls as a delta between snapshots, for each of the networks?

**Comment by runtian-zhou** [comment]
Right now the PR only did the generation from yaml path. In the next PR I can add the logic to fetch the config from rest endpoint and compare it against the yaml file. cc @igor-aptos

File: aptos-move/aptos-release-builder/src/components/mod.rs [hunk 8]
```diff
@@ -3,48 +3,68 @@
...
 }
 
 impl ReleaseConfig {
-    pub fn generate_release_proposal_scripts(
-        &self,
-        base_path: &Path,
-        is_testnet: bool,
-    ) -> Result<()> {
+    pub fn generate_release_proposal_scripts(&self, base_path: &Path) -> Result<()> {
         let mut result = vec![];
```
Comment (igor-aptos) [inline aptos-move/aptos-release-builder/src/components/mod.rs:36]: you are not returning the result, or doing anything else with it?

**Review by igor-aptos — APPROVED** [review]
makes sense, just wanted to understand where this fits in.

File: aptos-move/aptos-release-builder/src/components/mod.rs (hunk 8 above)
Comment (runtian-zhou) [reply aptos-move/aptos-release-builder/src/components/mod.rs:36]: This function will generate a set of move scripts into the `base_path`.

**Pull request merged**